core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1658,10 +1658,21 @@ class KafkaController(val config: KafkaConfig,
   }
 
   private def processTopicIds(topicIdAssignments: Set[TopicIdReplicaAssignment]): Unit = {
-    if (config.usesTopicId) {
-      val updated = zkClient.setTopicIds(topicIdAssignments.filter(_.topicId.isEmpty), controllerContext.epochZkVersion)
-      val allTopicIdAssignments = updated ++ topicIdAssignments.filter(_.topicId.isDefined)
-      allTopicIdAssignments.foreach(topicIdAssignment => controllerContext.addTopicId(topicIdAssignment.topic, topicIdAssignment.topicId.get))
+    // Create topic IDs for topics missing them if we are using topic IDs
+    // Otherwise, maintain what we have in the topicZNode
+    val updated = if (config.usesTopicId) {
+      zkClient.setTopicIds(topicIdAssignments.filter(_.topicId.isEmpty), controllerContext.epochZkVersion)
+    } else {
+      Set[TopicIdReplicaAssignment]()
+    }
+
+    // Add topic IDs to controller context
+    // If we don't have IBP 2.8, but are running 2.8 code, put any topic IDs from the ZNode in controller context
+    // This is to avoid losing topic IDs during operations like partition reassignments while the cluster is in a mixed state
+    (updated ++ topicIdAssignments).foreach { topicIdAssignment =>
```

Comment:
Is it correct to use `topicIdAssignments` here? It seems that `updated` will contain `TopicIdReplicaAssignment` twice for the partitions that have been assigned a new topic id above. I guess it works because those don't have a topic id defined but it is a bit confusing.

Instead, could we do something as follow:

```
val updatedTopicIdAssignments = if (config.usesTopicId) {
  val (withTopicIds, withoutTopicIds) = topicIdAssignments.partition(_.topicId.isDefined)
  withTopicIds ++ zkClient.setTopicIds(withoutTopicIds, controllerContext.epochZkVersion)
} else {
  topicIdAssignments
}

updatedTopicIdAssignments.foreach { topicIdAssignment =>
  ...
}
```

Reply:
We won't do it twice since topicIdAssignment.topicId is None for the partitions before they are assigned a topic ID. But I'll look at the way you suggested since it is clearer.